### PR TITLE
Support the siduck datagram test protocol in interop tools

### PR DESCRIPTION
**Pull request by Ralith**
See https://tools.ietf.org/html/draft-pardue-quic-siduck-00 for details.

### CODE REVIEW

File: interop/src/server.rs [hunk 1]
```diff
@@ -450,3 +451,32 @@ const HOME: &str = r##"
...
+                return Ok(());
+            }
+            Err(e) => {
+                return Err(e.into());
+            }
+            Ok(data) => {
+                if &data[..] == b"quack" {
+                    connection.send_datagram(b"quack-ack"[..].into())?;
+                } else {
+                    connection.close(0x101u32.into(), b"quack quack quack");
```
Comment (djc) [inline interop/src/server.rs:475]: Maybe name the magic integer value?

File: interop/src/main.rs [hunk 2]
```diff
@@ -165,9 +169,10 @@ async fn main() {
             port: opt.port.unwrap_or(4433),
             retry_port: opt.retry_port.unwrap_or(4434),
             sequential: opt.seq,
-            alpn: match (opt.h3, opt.hq) {
-                (false, true) => Alpn::Hq,
-                (true, false) => Alpn::H3,
+            alpn: match (opt.h3, opt.hq, opt.siduck) {
+                (false, true, _) => Alpn::Hq,
+                (true, false, _) => Alpn::H3,
+                (_, _, true) => Alpn::Siduck,
```
Comment (djc) [inline interop/src/main.rs:175]: I think we should get rid of the `_`s here, except for a fallthrough case that emits an error (or just panics).

File: interop/src/main.rs [hunk 3]
```diff
@@ -180,9 +185,10 @@ async fn main() {
             .collect();
 
         peers.iter_mut().for_each(|mut x| {
-            match (opt.h3, opt.hq) {
-                (false, true) => x.alpn = Alpn::Hq,
-                (true, false) => x.alpn = Alpn::H3,
+            match (opt.h3, opt.hq, opt.siduck) {
+                (false, true, _) => x.alpn = Alpn::Hq,
+                (true, false, _) => x.alpn = Alpn::H3,
+                (_, _, true) => x.alpn = Alpn::Siduck,
```
Comment (djc) [inline interop/src/main.rs:191]: This looks suspiciously close to the other one. DRY? (I know it's not new, but I'll invoke the boy scout rule.)

File: interop/src/main.rs [hunk 4]
```diff
@@ -219,10 +225,11 @@ async fn main() {
 async fn run(peer: &Peer, keylog: bool) -> Result<String> {
     let state = State::try_new(&peer, keylog)?;
     let result = match peer.alpn {
-        Alpn::Hq => state.run_hq().instrument(info_span!("hq")).await?,
-        _ => state.run_h3().instrument(info_span!("h3")).await?,
+        Alpn::Hq => state.run_hq().instrument(info_span!("hq")).await?.format(),
+        Alpn::Siduck => state.run_siduck().instrument(info_span!("siduck")).await?,
+        _ => state.run_h3().instrument(info_span!("h3")).await?.format(),
```
Comment (djc) [inline interop/src/main.rs:230]: Let's also get rid of the `_` in this case.

**Comment by Ralith** [comment]
Comments applied. Looks like azure pipelines is sticking its nose into our business again for some reason?

**Comment by djc** [comment]
That bit about Azure is weird. I've just deleted the Azure pipeline, so that should take care of business.

**Pull request merged**